### ISSUE TITLE
FSDP followup improvements

### ISSUE THREAD

**Issue by mrshenli** [pytorch/pytorch#64618]
We will keep adding items to this issue as we read #63881

- [ ] API redesign: #64394 
- [ ] Support multiple activation checkpoints in FSDP: #64633
- [ ] Refactor FSDP unit tests to use common distributed helpers: #64664
- [ ] FSDP forward return type should not be Tensor type: #64656
- [ ]  Some refactoring and documentation enhancements: #64685
- [ ] Set flat parameters as views to the flatten tensor and keep param names: #64464
- [ ] FSDP checkpoint APIs: #64625
- [ ] FSDP gradient ready order may not be the same among ranks: #64629
- [ ] Remove all the '.data' usage in FSDP repo: #64630
- [ ] Add a sharded grad scaler in PyTorch distributed: #64632
- [ ] FSDP initialization redesign: #64657
- [ ] Sporadic illegal memory access in FSDP when flatten_parameters = True: #64662
- [ ] reduce_scatter does not overlap with all_gather in backward pass in FSDP: #64803
- [ ] Stablize FSDP hook registration: #65194
- [ ] Detect globally unused params for FSDP: #65195
- [ ] Investigate unexpected stream execution in FSDP: #65299
- [ ] Bucket redesign in FSDP: #65349
- [ ] Have a generic no_sync for all synchronized training features: #64684
- [ ] No shard + CPU offload in FSDP: #64772
- [ ] Sometimes CudaMalloc is slow in FSDP: #65078
- [ ] Support optimizer overlap with backward pass in FSDP: #64661
- [ ] Avoid H2D copy and stream sync in clip_grad_norm_: #65110
- [ ] Explicit prefetching data in FSDP: #65248
- [ ] Support dividing the `flat_parameter` into smaller ones, and only allgather individual smaller ones on-demand (this should help save peak memory consumption). #64686
- [ ] Add compression hooks for FSDP: #64665
- [ ] Add FSDP logging data: #64667
- [ ] Improve FSDP debug ability: #64669
- [ ] Extend generic join to support FSDP: #64683
- [ ] Define finer grain TrainingState in FSDP: #64674
- [ ] Consolidate DDPSink with FSDP pre-backward hooks: #65193
- [ ] calc_grad_norm_ always compute the norm in full precision: #65108
- [ ] avoid to append a list of chunks in _get_shard(): #65111
- [ ] remove torch.cuda.synchronize() in state_dict() : #65117
- [ ] https://github.com/pytorch/pytorch/issues/67288
- [ ] #66908
- [ ] #66919
- [ ] #66911
- [ ] https://github.com/pytorch/pytorch/issues/67714

**Comment by mrshenli** [comment]
cc @zhaojuanmao

**Comment by zhaojuanmao** [comment]
@mrshenli thanks a lot for adding all these comments and thoughts, I added more issues and merged some of your comments into those issues.  

Basically [Beta] tag mostly means they are more high pri, let us continue to add comments and thoughts into this list!

**Comment by zhaojuanmao** [comment]
@anj-s @min-xu-ai

**Comment by min-xu-ai** [comment]
Thanks for tagging! I think there is also a perf regression still being investigated by Liang and his team.

Be sure to include the recent fixes and enhances from fairscale repo as well. Thanks!

**Comment by zhaojuanmao** [comment]
add [prototype] tag to indicate work items need to be done for prototype release

**Comment by zhaojuanmao** [comment]
closing this, each individual issue with label "module: fsdp" is tracked
